File: menger-sponge-fractal/Box.js
class Box {
  constructor(x, y, z, r) {
    this.position = createVector(x, y, z);
    this.size = r;
  }
  
  show () {
    // initilize a new drawing state
    push();
    
    // set coordinates
    translate(this.position.x, this.position.y, this.position.z);
    
    // draw the box at that position
    fill(150, 0, 255);
    stroke(50, 0, 200, 50);
    strokeWeight(1 + this.size / 40);
    box(this.size);
    
    // Restore original state
    pop();
  }
  
  generate () {
    const boxes = [];
    
    for (let i = -1; i < 2; i++) {
      for (let j = -1; j < 2; j++) {
        for (let k = -1; k < 2; k++) {
          const sum = abs(i) + abs(j) + abs(k);
          
          // write > 1 to display as box
          // write <= 1 to display as snowflake
          if (sum <= 1) {
            // new box's dimensions should be the third of the current one
            const newSize = this.size / 3;
          
            // offset the new coordinates
            const x = this.position.x + i * newSize;
            const y = this.position.y + j * newSize;
            const z = this.position.z + k * newSize;
          
            const newBox = new Box(x, y, z, newSize);
          
            boxes.push(newBox);
          }
        }
      }
    }
    
    return boxes;
  }
}

File: sketches/menger_sponge_fractal/Box.py
class Box():
    def __init__(self, x, y, z, r):
        self.position = PVector(x, y, z)
        self.size = r
    
    
    def show(self):
        push()
        translate(self.position.x, self.position.y, self.position.z)
        fill(150, 0, 255)
        stroke(50, 0, 200, 50)
        strokeWeight(1 + self.size / 40)
        box(self.size)
        pop()
        
        
    def generate(self):
        boxes = []
        
        for i in range(-1, 2):
            for j in range(-1, 2):
                for k in range(-1, 2):
                    sum = abs(i) + abs(j) + abs(k)
                    
                    # write > 1 to display as classic sponge
                    # write <= 1 to display as opposite
                    if sum > 1:
                        new_size = self.size / 2.5
                        
                        x = self.position.x + i * new_size
                        y = self.position.y + j * new_size
                        z = self.position.z + k * new_size
                        
                        new_box = Box(x, y, z, new_size)
                        boxes.append(new_box)
                        
        return boxes
    
